src/Microsoft.CodeQuality.Analyzers/UnitTests/ApiDesignGuidelines/PropagateCancellationTokensWhenPossibleTests.cs

```diff
@@ -4,11 +4,266 @@
+        public void WipCSharp_DefaultCancellationTokenPassed()
+        {
+            var source = @"
+using System.Threading;
+using System.Threading.Tasks;
+
+class C
+{
+    Task M1(CancellationToken p1) => throw null;
+
+    Task M2(CancellationToken p1) => M1(default(CancellationToken));
+}
+";
+            VerifyCSharp(source, GetCSharpResultAt(9, 38));
+        }
+
+        [Fact]
+        public void WipCSharp_CancellationTokenNonePassed()
+        {
+            var source = @"
+using System.Threading;
+using System.Threading.Tasks;
+
+class C
+{
+    Task M1(CancellationToken p1) => throw null;
+
+    Task M2(CancellationToken p1) => M1(CancellationToken.None);
+}
+";
+            VerifyCSharp(source, GetCSharpResultAt(9, 38));
+        }
+
+        [Fact]
+        public void WipCSharp_CancellationTokenInScope_InstanceField()
+        {
+            var source = @"
+using System.Threading;
+using System.Threading.Tasks;
+
+class C
+{
+    private CancellationToken _f1;
+
+    Task M1() => throw null;
+
+    Task M1(CancellationToken p1) => throw null;
+
+    Task M2() => M1();
+}
+";
+            VerifyCSharp(source, GetCSharpResultAt(13, 18));
+        }
+
+        [Fact]
+        public void WipCSharp_CancellationTokenInScope_StaticField()
+        {
+            var source = @"
+using System.Threading;
+using System.Threading.Tasks;
+
+class C
+{
+    private static CancellationToken _s1;
+
+    Task M1() => throw null;
+
+    Task M1(CancellationToken p1) => throw null;
+
+    Task M2() => M1();
+}
+";
+            VerifyCSharp(source, GetCSharpResultAt(13, 18));
+        }
+
+        [Fact]
+        public void WipCSharp_CancellationTokenInScope_LocalVariable()
+        {
+            var source = @"
+using System.Threading;
+using System.Threading.Tasks;
+
+class C
+{
+    Task M1() => throw null;
+
+    Task M1(CancellationToken p1) => throw null;
+
+    Task M2()
+    {
+        var l1 = CancellationToken.None; // No dataflow
+        return M1();
+    }
+}
+";
+            VerifyCSharp(source, GetCSharpResultAt(14, 16));
+        }
+
+        [Fact]
+        public void WipCSharp_CancellationTokenIsPropagated_NoDiagnostics()
+        {
+            var source = @"
+using System.Threading;
+using System.Threading.Tasks;
+
+class C
+{
+    Task M1() => throw null;
+
+    Task M1(CancellationToken p1) => throw null;
+
+    Task M2(CancellationToken p1) => M1(p1);
+}
+";
+            VerifyCSharp(source);
+        }
+
+        [Fact]
+        public void WipCSharp_NoCancellationTokenOverloadAvailable_NoDiagnostics()
+        {
+            var source = @"
+using System.Threading;
+using System.Threading.Tasks;
+
+class C
+{
+    Task M1() => throw null;
+
+    Task M2(CancellationToken p1) => M1();
+}
+";
+            VerifyCSharp(source);
+        }
+
+        [Fact(Skip = "https://github.com/dotnet/roslyn-analyzers/issues/1476")]
+        public void WipCSharp_CancellationTokenInScope_UninitializedLocalVariable_NoDiagnostics()
```

Comment:
This scenario definitely looks like it will need dataflow analysis.

Reply:
See previous comment (couldn't figure out how to use `DataFlowAnalysis`)